datasource/json-directory.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- * A simple synchronised datasource backed by JSON files in directories.
+--
+module Main where
+
+import Control.Exception
+import Data.Aeson
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Lazy as BL
+import Data.Monoid
+import qualified Data.Text as T
+import Options.Applicative
+import System.Directory
+import System.Exit
+import System.FilePath.Posix
+import System.Posix.Files
+import System.Random
+
+import Synchronise.DataSource
+import Synchronise.Document
+import Synchronise.Identifier
+
+
+data JSONOpts = JSONOpts
+  { baseDir :: FilePath
+  , cmmnd   :: JSONCommand }
+
+data JSONCommand
+  = Create
+  | Update FilePath
+  | Read   FilePath
+  | Delete FilePath
+
+pOpts :: Parser JSONOpts
+pOpts
+  =   JSONOpts
+  <$> strArgument (metavar "BASE")
+  <*> pCommand
+
+pCommand :: Parser JSONCommand
+pCommand
+  = subparser
+  (  command "create" (info pCreate (progDesc "Create a new unique JSON filename"))
+  <> command "update" (info pUpdate (progDesc "Writes a document to the given JSON directory"))
+  <> command "read"   (info pRead   (progDesc "Reads a document in the given JSON directory"))
+  <> command "delete" (info pDelete (progDesc "Deletes a document in the given JSON directory")))
+  where pCreate = pure Create
+        pUpdate = Update <$> strArgument (metavar "FILEPATH")
+        pRead   = Read   <$> strArgument (metavar "FILEPATH")
+        pDelete = Delete <$> strArgument (metavar "FILEPATH")
+
+main :: IO ()
+main = do
+  JSONOpts{..} <- execParser toplevel
+  createDirectoryIfMissing True baseDir
+  exit <- case cmmnd of
+    Create   -> jsonCreate baseDir
+    Read   f -> jsonRead   baseDir f
+    Update f -> jsonUpdate baseDir f
+    Delete f -> jsonDelete baseDir f
+  exitWith exit
+  where toplevel = info pOpts mempty
+
+--------------------------------------------------------------------------------
+
+-- | Creates a new JSON file under the path specified.
+--   Prints the filename to `stdout`.
+--
+jsonCreate :: FilePath -> IO ExitCode
+jsonCreate base = do
+  k      <-  T.pack . take 64 . randomRs ('a', 'z')
+         <$> newStdGen
+  exists <- fileExist (mkJSONFilename base k)
+  if   exists
+  then jsonCreate base
+  else do print k
```

Comment:
no? by creating you only get a "primary key" which will get written when you update.